_studio/mfx_lib/encode_hw/h264/src/mfx_h264_encode_hw_utils_new.cpp

```diff
@@ -2516,19 +2535,28 @@ mfxStatus MfxHwH264Encode::CodeAsSkipFrame(     VideoCORE &            core,
     else
     {
         mfxI32 fid = task.m_fid[0];
-        MFX_CHECK(task.m_list0[fid].Size(), MFX_ERR_UNDEFINED_BEHAVIOR);
-        DpbFrame& refFrame = task.m_dpb[0][task.m_list0[fid][0] & 127];
+        mfxU8 ind = ((task.GetFrameType() & MFX_FRAMETYPE_B)  && task.m_list1[fid].Size() != 0) ? task.m_list1[fid][0] : task.m_list0[fid][0];
+        MFX_CHECK(ind < 15, MFX_ERR_UNDEFINED_BEHAVIOR);
+
+        DpbFrame& refFrame = task.m_dpb[0][ind];
         mfxFrameData curr = {};
         mfxFrameData ref = {};
         curr.MemId = task.m_midRaw;
         ref.MemId  = refFrame.m_midRec;
 
         mfxFrameSurface1 surfSrc = { {0,}, video.mfx.FrameInfo, ref  };
         mfxFrameSurface1 surfDst = { {0,}, video.mfx.FrameInfo, curr };
-        sts = core.DoFastCopyWrapper(&surfDst,MFX_MEMTYPE_INTERNAL_FRAME|MFX_MEMTYPE_DXVA2_DECODER_TARGET|MFX_MEMTYPE_FROM_ENCODE, &surfSrc, MFX_MEMTYPE_INTERNAL_FRAME|MFX_MEMTYPE_DXVA2_DECODER_TARGET|MFX_MEMTYPE_FROM_ENCODE);
+        if ((poolRec.GetFlag(refFrame.m_frameIdx) & H264_FRAME_FLAG_READY) != 0)
+        {
+            sts = core.DoFastCopyWrapper(&surfDst, MFX_MEMTYPE_INTERNAL_FRAME | MFX_MEMTYPE_DXVA2_DECODER_TARGET | MFX_MEMTYPE_FROM_ENCODE, &surfSrc, MFX_MEMTYPE_INTERNAL_FRAME | MFX_MEMTYPE_DXVA2_DECODER_TARGET | MFX_MEMTYPE_FROM_ENCODE);
 
+            if (ind != 0)
+                poolRec.SetFlag(refFrame.m_frameIdx, H264_FRAME_FLAG_SKIPPED);
+        }
+        else
```

Comment:
```if```\```else``` constructions, where some of the branches is pure ```return```` statement, always can be rewritten without ```else``` branch with ```return``` statement covered by ```if```. It often leads to readability improvement. In this case it would be
```
if ((poolRec.GetFlag(refFrame.m_frameIdx) & H264_FRAME_FLAG_READY) == 0)
   return MFX_WRN_DEVICE_BUSY;

sts = core.DoFastCopyWrapper(&surfDst, MFX_MEMTYPE_INTERNAL_FRAME | MFX_MEMTYPE_DXVA2_DECODER_TARGET | MFX_MEMTYPE_FROM_ENCODE, &surfSrc, MFX_MEMTYPE_INTERNAL_FRAME | MFX_MEMTYPE_DXVA2_DECODER_TARGET | MFX_MEMTYPE_FROM_ENCODE);

if (ind != 0)
    poolRec.SetFlag(refFrame.m_frameIdx, H264_FRAME_FLAG_SKIPPED);
```